lib/python/rosie/ws.py

```diff
@@ -99,72 +168,56 @@ def __init__(self, props, prefix, db_url):
         if source_url_node is not None:
             self.source_url = source_url_node.value
         self.dao = rosie.db.DAO(db_url)
+        self.root_service_url = root_service_url
+
+        # Arguments for API points and formatting
+        self.q_args = self.get_arguments("q")
+        self.s_args = self.get_arguments("s")
+        # Need to provide a default, else get MissingArgumentError w/ no arg:
+        self.format = self.get_argument("format", default=None)
 
     def __call__(self):
         """Dummy."""
         pass
 
-    @cherrypy.expose
-    def index(self, *_):
+    # Decorator to ensure there is a trailing slash since buttons for keys
+    # otherwise go to wrong URLs for "/rosie/key" (e.g. -> "rosie/query?...")
+    @tornado.web.addslash
+    def get(self, *_):
         """Provide the index page."""
         try:
-            return self._render()
+            self._render()
         except (KeyError, AttributeError, jinja2.exceptions.TemplateError):
             import traceback
             traceback.print_exc()
         except rosie.db.RosieDatabaseConnectError as exc:
-            raise cherrypy.HTTPError(404, str(exc))
+            raise tornado.web.HTTPError(404, str(exc))
 
-    @cherrypy.expose
     def hello(self, format=None):
         """Say Hello on success."""
-        if cherrypy.request.login:
-            data = self.HELLO % cherrypy.request.login
+        if self.get_current_user():
+            data = self.HELLO % self.get_current_user()
         else:
             data = self.HELLO % "user"
-        if format == "json":
-            return json.dumps(data)
+        if self.format == "json":
+            self.write(json.dumps(data))
+            return
         return data
```

Comment:
Out of interest, how come we write JSON data (and return None) but return strings?

Reply:
And also... `format` is an argument of the REST API. Not sure why it has become a class variable.

Reply:
For my replies to both these questions, relating intrinsically to the nature of Tornado, I found nice code examples for simple apps [under an official 'demos' repo, e.g. for a 'blog' example](https://github.com/tornadoweb/tornado/blob/stable/demos/blog/blog.py) that you may find it useful to reference.

******

> how come we write JSON data (and return None) but return strings?

If JSON is requested by query, the data gets written in JSON to the output buffer, but a ``return`` is needed to bypass the ``return data`` line, otherwise both JSON & the rendered web page are served. This is not a the best way to go about that logically, but in the interest of minimising the changeset line count & logical flow I avoided recasting.

We don't actually return strings; this was an omission for the ``hello()`` method (thanks for highlighting it for me!) Data should simply be called as the argument to ``RequestHandler`` output methods such as ``self.write``, without any logical need to call ``return``, instead of returned, as for the JSON & as in the other methods rending the web pages:

https://github.com/metomi/rose/blob/98325edeff3ca0126cf6ce5f3d800dc038a03b9d/lib/python/rosie/ws.py#L232

> format is an argument of the REST API. Not sure why it has become a class variable.

In Tornado, as described further & here [quoted from the docs](https://www.tornadoweb.org/en/stable/guide/structure.html#handling-request-input), "request data in the formats used by HTML forms will be parsed for you and is made available in methods like ``get_query_argument`` and ``get_body_argument``" or as implemented here via [the Request Handler method covering both query & body args](https://www.tornadoweb.org/en/stable/web.html?highlight=get_argument#tornado.web.RequestHandler.get_argument):

https://github.com/metomi/rose/blob/98325edeff3ca0126cf6ce5f3d800dc038a03b9d/lib/python/rosie/ws.py#L177

where I assigned it to an instance variable to avoid having to call the method several times.

The ``format`` argument is not recognised as a general REST API key by keyword, as in ``cherrypy``, & needs to be fetched via such methods. You could see this via attempting to revert back to ``if format == "json":`` where it does not get recognised, brekaing the functionality.

Does that make sense? Also, if you can see a better way to process this & perhaps the other arguments, I am happy to try a new approach.

Reply:
Sorry @matthewrmshin, @oliver-sanders just elaborated on what more specifically you were referring to there in your comment. You are right, the class variable nature of some of these API arguments is inappropriate & I now know is the cause of an issue of persistent query arguments that I had been seeing & not previously understanding. Thanks both, this comment & discussion has really helped with my understanding.